src/main/kotlin/com/refinedmods/refinedstorage/apiimpl/network/NetworkNodeGraph.kt

```diff
@@ -0,0 +1,140 @@
+    private var invalidating = false
+    override fun invalidate(action: Action, world: World, origin: BlockPos) {
+        invalidating = true
+        val operator = Operator(action)
+        val originNode: INetworkNode? = NetworkUtils.getNodeFromBlockEntity(world.getBlockEntity(origin))
+        if (originNode is INetworkNodeVisitor) {
+            (originNode as INetworkNodeVisitor).visit(operator)
+        }
+        var currentVisitor: Visitor
+        while (operator.toCheck.poll().also { currentVisitor = it } != null) {
+            currentVisitor.visit(operator)
+        }
+        nodes = operator.foundNodes
+        if (action == Action.PERFORM) {
+            for (node in operator.newNodes) {
+                node.onConnected(network)
+            }
+            for (node in operator.previousNodes) {
+                node.onDisconnected(network)
+            }
+            actions.forEach { h: Consumer<INetwork> -> h.accept(network) }
+            actions.clear()
+            if (operator.newNodes.isNotEmpty() || operator.previousNodes.isNotEmpty()) {
+                listeners.forEach(Consumer<INetworkNodeGraphListener> { obj: INetworkNodeGraphListener -> obj.onChanged() })
+            }
+        }
+        invalidating = false
+    }
+
+    override fun runActionWhenPossible(handler: Consumer<INetwork>) {
+        when {
+            invalidating -> actions.add(handler)
+            else -> handler.accept(network)
+        }
+    }
+
+    override fun all(): Collection<INetworkNode> {
+        return nodes
+    }
+
+    override fun addListener(listener: INetworkNodeGraphListener) {
+        listeners.add(listener)
+    }
+
+    override fun disconnectAll() {
+        nodes.forEach(Consumer<INetworkNode> { n: INetworkNode -> n.onDisconnected(network) })
+        nodes.clear()
+        listeners.forEach(Consumer<INetworkNodeGraphListener> { obj: INetworkNodeGraphListener -> obj.onChanged() })
+    }
+
+    val world: World
+        get() = network.world
+
+    private fun dropConflictingBlock(world: World, pos: BlockPos) {
+        if (network.position != pos) {
+            Block.dropStacks(world.getBlockState(pos), world, pos, world.getBlockEntity(pos))
+            world.removeBlock(pos, false)
+        }
+    }
+
+    private inner class Operator(override val action: Action) : INetworkNodeVisitor.Operator {
+        val foundNodes: MutableSet<INetworkNode> = Sets.newConcurrentHashSet() // All scanned nodes
+        val newNodes: MutableSet<INetworkNode> = Sets.newConcurrentHashSet<INetworkNode>() // All scanned new nodes, that didn't appear in the list before
+        val previousNodes: MutableSet<INetworkNode> = Sets.newConcurrentHashSet(nodes) // All unscanned nodes (nodes that were in the previous list, but not in the new list)
+        val toCheck: Queue<Visitor> = ArrayDeque()
+        override fun apply(world: World, pos: BlockPos, side: Direction) {
+            world.getBlockEntity(pos)?.let{tile ->
```

Comment:
I prefer
```kotlin
val tile = world.getBlockEntity(pos)?: return
```

Reply:
Do you prefer that in all instances of `let`? Or is there a reason for it in this instance?